xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -684,146 +692,164 @@ protected void handleLdsResponse(String versionInfo, List<Any> resources, String
+      } catch (InvalidProtocolBufferException e) {
+        errors.add("RDS response Resource index " + i + " - can't decode RouteConfiguration: " + e);
+        continue;
       }
-    } catch (InvalidProtocolBufferException e) {
-      getLogger().log(
-          XdsLogLevel.WARNING, "Failed to unpack RouteConfiguration in RDS response {0}", e);
-      nackResponse(ResourceType.RDS, nonce, "Malformed RDS response: " + e);
-      return;
-    }
-    getLogger().log(
-        XdsLogLevel.INFO, "Received RDS response for resources: {0}", routeConfigs.keySet());
+      String routeConfigName = routeConfig.getName();
+      unpackedResources.add(routeConfigName);
 
-    Map<String, RdsUpdate> rdsUpdates = new HashMap<>();
-    boolean parseFilter = enableFaultInjection && isResourceV3;
-    for (Map.Entry<String, RouteConfiguration> entry : routeConfigs.entrySet()) {
-      String routeConfigName = entry.getKey();
-      RouteConfiguration routeConfig = entry.getValue();
-      List<VirtualHost> virtualHosts =
-          new ArrayList<>(routeConfig.getVirtualHostsCount());
-      for (io.envoyproxy.envoy.config.route.v3.VirtualHost virtualHostProto
-          : routeConfig.getVirtualHostsList()) {
-        StructOrError<VirtualHost> virtualHost = parseVirtualHost(virtualHostProto, parseFilter);
-        if (virtualHost.getErrorDetail() != null) {
-          nackResponse(ResourceType.RDS, nonce, "RouteConfiguration " + routeConfigName
-              + " contains invalid virtual host: " + virtualHost.getErrorDetail());
-          return;
-        }
-        virtualHosts.add(virtualHost.getStruct());
+      // Process RouteConfiguration into RdsUpdate.
+      RdsUpdate rdsUpdate;
+      boolean isResourceV3 = resource.getTypeUrl().equals(ResourceType.RDS.typeUrl());
+      try {
+        rdsUpdate = processRouteConfiguration(routeConfig, enableFaultInjection && isResourceV3);
+      } catch (ResourceInvalidException e) {
+        errors.add(
+            "RDS response RouteConfiguration '" + routeConfigName + "' validation error: " + e
+                .getMessage());
+        continue;
       }
-      rdsUpdates.put(routeConfigName, new RdsUpdate(virtualHosts));
+
+      parsedResources.put(routeConfigName, new ParsedResource(rdsUpdate, resource));
     }
-    ackResponse(ResourceType.RDS, versionInfo, nonce);
+    getLogger().log(XdsLogLevel.INFO,
+        "Received RDS Response version {0} nonce {1}. Parsed resources: {2}",
+        versionInfo, nonce, unpackedResources);
 
-    for (String resource : rdsResourceSubscribers.keySet()) {
-      if (rdsUpdates.containsKey(resource)) {
-        ResourceSubscriber subscriber = rdsResourceSubscribers.get(resource);
-        subscriber.onData(rdsUpdates.get(resource));
+    if (!errors.isEmpty()) {
+      handleResourcesNacked(ResourceType.RDS, unpackedResources, versionInfo, nonce, errors);
+    } else {
+      handleResourcesAcked(ResourceType.RDS, parsedResources, versionInfo, nonce, false);
```

Comment:
Should be `true`.

Reply:
Wait, why should this be `true`? This is an RDS response, which is incremental. Anything not included in this response may come right after, so we should not conclude them being absent at this point.

Reply:
You are right. I misremembered.